src/binary/writer.rs

```diff
@@ -0,0 +1,1324 @@
+    fn encode_annotated_scalar(&mut self,
+                               mut scalar_write_fn: impl FnMut(&mut Vec<u8>) -> IonResult<()>) -> IonResult<()> {
+
+        // Encode the scalar into the buffer, but do not push the IoRange yet.
+        let value_io_range: IoRange = self.encode_to_buffer(|writer| {
+            scalar_write_fn(&mut writer.buffer)
+        })?;
+
+        // Create ranges that will ultimately point to the encoded components of the annotations
+        // wrapper for the value.
+        let mut header_io_range: Range<usize> = 0..0;
+        let mut annotations_seq_length_io_range: Range<usize> = 0..0;
+        let mut annotations_seq_io_range: Range<usize> = 0..0;
+
+        // Using the encoded length of the value, encode the annotations wrapper and populate
+        // the IO ranges we created above.
+        self.encode_annotation_wrapper(
+            &mut header_io_range,
+            &mut annotations_seq_length_io_range,
+            &mut annotations_seq_io_range,
+            value_io_range.len(),
+        )?;
+
+        // Push the IO ranges in the correct order so the encoded bytes will be written in the
+        // correct order when the user calls `flush()`.
+        self.io_ranges.extend_from_slice(&[
+            header_io_range,
+            annotations_seq_length_io_range,
+            annotations_seq_io_range,
+            value_io_range
+        ]);
+
+        self.push_empty_io_range();
+
+        Ok(())
+    }
+
+    // Writes the annotations wrapper for a value of a given length. Callers should encode the
+    // value to the buffer, call this method with the value's encoded length, then push
+    // the populated IoRanges in the necessary order.
+    fn encode_annotation_wrapper(&mut self,
+                                 header_io_range: &mut IoRange,
+                                 annotations_seq_length_io_range: &mut IoRange,
+                                 annotations_seq_io_range: &mut IoRange,
+                                 wrapped_value_length: usize) -> IonResult<()> {
+        // Encode the sequence of annotations and make a note of the encoded length.
+        // The return value of mem::replace is the original range value, which is always 0..0.
+        // We can safely ignore it.
+        let _ = mem::replace(annotations_seq_io_range, self.encode_to_buffer(|writer| {
+            let range = writer.current_value_annotations_range();
+            let annotations = &writer.annotations_all_levels[range];
+            for annotation_id in annotations {
+                VarUInt::write_u64(&mut writer.buffer, *annotation_id as u64)?;
+            }
+            Ok(())
+        })?);
+        let annotation_sequence_encoded_length = annotations_seq_io_range.len();
+
+        // Encode the length of the annotations sequence as a VarUInt.
+        let _ = mem::replace(annotations_seq_length_io_range, self.encode_to_buffer(|writer| {
+            let _num_bytes = VarUInt::write_u64(
```

Comment:
do we need `_num_bytes`. It looks that is it not used? Replace with `_`?

Reply:
I'm wary of using `_` by itself too often, because as a reader visiting new code it usually means I have to fire up an IDE to get hints about what's being discarded. I still do it sometimes, including just above this code snippet; [the first time I wrote an explanation for why we could ignore it](https://github.com/amzn/ion-rust/blob/f246c0a6189ee8065a537299800eca24c204a7c8/src/binary/writer.rs#L275-L277), but didn't add the same disclaimer elsewhere.

Prefixing the variable name with `_` is enough to satisfy the compiler/style guidelines as it flags it as intentionally unused. I happened to know what the value I was ignoring was, so I used a meaningful variable name instead of adding a comment.

```
// Ignore the number of bytes that were written.
let _ = VarUInt::write_u64(...);
// vs. 
let _num_bytes = VarUInt::write_u64(...);
```

I'm unsure if there's a formal style guideline around this; if so, I'm happy to defer to it.